x/nft-transfer/keeper/packet.go

```diff
@@ -0,0 +1,198 @@
+		sourceChannel, fullClassPath)
+
+	for _, tokenID := range tokenIDs {
+		nft, err := k.nftKeeper.GetNFT(ctx, classID, tokenID)
+		if err != nil {
+			return channeltypes.Packet{}, err
+		}
+		tokenURIs = append(tokenURIs, nft.GetURI())
+
+		owner := nft.GetOwner()
+		if !sender.Equals(owner) {
+			return channeltypes.Packet{}, sdkerrors.Wrap(sdkerrors.ErrUnauthorized, "not token owner")
+		}
+
+		if isAwayFromOrigin {
+			// create the escrow address for the tokens
+			escrowAddress := types.GetEscrowAddress(sourcePort, sourceChannel)
+			if err := k.nftKeeper.TransferOwner(ctx, classID, tokenID, sender, escrowAddress); err != nil {
+				return channeltypes.Packet{}, err
+			}
+		} else {
+			if err := k.nftKeeper.BurnNFTUnverified(ctx, classID, tokenID, sender); err != nil {
+				return channeltypes.Packet{}, err
+			}
+		}
+	}
+
+	packetData := types.NewNonFungibleTokenPacketData(
+		fullClassPath, denom.Uri, tokenIDs, tokenURIs, sender.String(), receiver,
+	)
+
+	return channeltypes.NewPacket(
+		packetData.GetBytes(),
+		sequence,
+		sourcePort,
+		sourceChannel,
+		destinationPort,
+		destinationChannel,
+		timeoutHeight,
+		timeoutTimestamp,
+	), nil
+}
+
+// processReceivedPacket will mint the tokens to receiver account
+// if the token was away from origin chain . Otherwise, the sent tokens
+// were burnt in the sending chain and will unescrow the token to receiver
+// in the destination chain
+func (k Keeper) processReceivedPacket(ctx sdk.Context, packet channeltypes.Packet,
+	data types.NonFungibleTokenPacketData,
+) error {
+	receiver, err := sdk.AccAddressFromBech32(data.Receiver)
+	if err != nil {
+		return err
+	}
+
+	isAwayFromOrigin := types.IsAwayFromOrigin(packet.GetSourcePort(), packet.GetSourceChannel(), data.ClassId)
+
+	// create the escrow address for creating denom and minting nft
+	escrowAddress := types.GetEscrowAddress(packet.GetDestPort(), packet.GetDestChannel())
+
+	if isAwayFromOrigin {
+		// since SendPacket did not prefix the classID, we must prefix classID here
+		classPrefix := types.GetClassPrefix(packet.GetDestPort(), packet.GetDestChannel())
+		// NOTE: sourcePrefix contains the trailing "/"
+		prefixedClassID := classPrefix + data.ClassId
+
+		// construct the class trace from the full raw classID
+		classTrace := types.ParseClassTrace(prefixedClassID)
+		if !k.HasClassTrace(ctx, classTrace.Hash()) {
+			k.SetClassTrace(ctx, classTrace)
+		}
+
+		voucherClassID := classTrace.IBCClassID()
+
+		if !k.nftKeeper.HasDenomID(ctx, voucherClassID) {
+			if err := k.nftKeeper.IssueDenom(ctx, voucherClassID, "", "", data.ClassUri, escrowAddress); err != nil {
+				return err
+			}
+		}
+
+		ctx.EventManager().EmitEvent(
+			sdk.NewEvent(
+				types.EventTypeClassTrace,
+				sdk.NewAttribute(types.AttributeKeyTraceHash, classTrace.Hash().String()),
+				sdk.NewAttribute(types.AttributeKeyClassID, voucherClassID),
+			),
+		)
+
+		for i, tokenID := range data.TokenIds {
```

Comment:
https://github.com/cosmos/ibc/tree/main/spec/app/ics-721-nft-transfer#packet-relay
has that loop outer, i.e. it checks class for every tokenid -- not sure if significant... I assume each transfer will only transfer nfts for the same denom/class?

Reply:
Yes. One message will only contain NFTs of a single denom. Here, we check if the denom exists (and create if it doesn't) before entering the loop.